File: frontend/types/importacionFormacion.ts
/**
 * Tipos del import de Formación por Excel. Espejan `schemas/importacion_formacion.py`.
 *
 * Archivo propio y no dentro de `types/importacionObjetivos.ts` por la misma razón que del lado
 * del backend: son dos vocabularios distintos y mezclarlos obliga a leer el doble para entender
 * uno. El contrato de fondo sí es el mismo — el preview devuelve la fila YA RESUELTA y el
 * confirmar la recibe de vuelta: resolver para MOSTRAR, revalidar para ESCRIBIR.
 */

/** Una fila que NO se carga, con el motivo ya redactado por el backend para el usuario. */
export interface FilaFormacionError {
  fila: number
  identificador: string
  motivo: string
}

export interface FilaFormacionPreview {
  fila: number
  titulo: string
  /** Lo que decía la celda, para que el usuario reconozca su propia planilla. */
  colaborador: string
  /**
   * 🔴 UNO DE LOS DOS, NUNCA NINGUNO: o la persona matcheó contra el padrón (`empleado_id` +
   * `empleado_nombre`, el nombre del padrón para que se vea QUÉ matcheó) o entra como texto
   * suelto (`nombre_libre`). No matchear NO es un error: la formación queda cargada igual y la
   * persona se puede vincular después.
   */
  empleado_id: string | null
  empleado_nombre: string | null
  nombre_libre: string | null
  estado: "pendiente" | "en_curso" | "completado"
  fecha_asignacion: string | null
  fecha_completado: string | null
  proyecto: string | null
  anio: string | null
  mes: string | null
  /** Atributos del catálogo que esta fila aporta si su capacitación hay que crearla. */
  tipo: string | null
  entidad_capacitadora: string | null
  modalidad: string | null
  duracion_horas: number | null
  avisos: string[]
}

/** Un título que no existe en el catálogo de la empresa y el confirmar va a crear. */
export interface CapacitacionACrear {
  nombre: string
  tipo: string | null
  entidad_capacitadora: string | null
  modalidad: string | null
  duracion_horas: number | null
  avisos: string[]
}

/**
 * Dos nombres crudos distintos que probablemente sean la misma persona.
 *
 * 🔴 EL SISTEMA NO LOS UNIFICA Y NO DEBE: decide RRHH. Por eso viajan al front en vez de
 * resolverse en el backend — es el único grupo del preview que pide una acción HUMANA sobre el
 * archivo antes de importar.
 */
export interface ParParecido {
  nombre_a: string
  nombre_b: string
  motivo: string
}

export interface ImportacionFormacionPreview {
  filas_validas: FilaFormacionPreview[]
  errores: FilaFormacionError[]
  capacitaciones_a_crear: CapacitacionACrear[]
  /** Nombres que van a entrar como `nombre_libre`, sin vincular a un legajo. */
  sin_match: string[]
  pares_parecidos: ParParecido[]
  hoja_leida: string | null
  total_hojas: number
}

/**
 * 🔴 El resultado NO es binario. El lote no aborta por una fila con problemas: se cargan las
 * válidas y se reportan las otras, así que `importados > 0` y `errores.length > 0` conviven.
 */
export interface ImportacionFormacionResultado {
  importados: number
  errores: FilaFormacionError[]
  /** Nombres creados en el catálogo por este lote. */
  capacitaciones_creadas: string[]
}

File: backend/schemas/importacion_formacion.py
"""
Schemas del import de Formación por Excel (preview → confirmar).

Molde: `schemas/importacion_objetivos.py`. Mismo contrato de fondo: el preview devuelve la fila
YA RESUELTA (empleado matcheado o nombre_libre, estado traducido, fechas derivadas) y el
confirmar recibe esas filas de vuelta — resolver para MOSTRAR, revalidar para ESCRIBIR, porque
el body viaja por la red y el cliente lo puede alterar.

🔴 `estado` es Literal y no str: la traducción del Excel ("Finalizado"→completado) ya ocurrió en
el preview, así que si al confirmar llega otra cosa es el CLIENTE alterando el body — eso muere
en el 422, no entra a chocar contra el CHECK de la base. Mismo criterio que el `tipo` de
objetivos.

🔴 La empresa viaja en el BODY del confirmar Y como Form en el preview, nunca en el header:
importar es una ACCIÓN y la empresa es un dato del formulario (Vista vs Acción). El Excel no
menciona sociedades — todo el lote va a la empresa elegida (decisión 1 del 19/8).
"""
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel


class FilaFormacionError(BaseModel):
    """Una fila que NO se carga, con el motivo en texto para quien tiene el Excel abierto."""
    fila: int
    identificador: str
    motivo: str


class FilaFormacionPreview(BaseModel):
    fila: int
    titulo: str
    colaborador: str                      # lo que decía la celda, para que el usuario lo reconozca
    # Uno de los dos, nunca ninguno: o matcheó contra el padrón o entra como nombre suelto.
    empleado_id: Optional[UUID] = None
    empleado_nombre: Optional[str] = None  # el nombre del padrón, para que se vea QUÉ matcheó
    nombre_libre: Optional[str] = None
    estado: Literal["pendiente", "en_curso", "completado"]
    fecha_asignacion: Optional[str] = None   # "YYYY-MM-DD", primer día del mes derivado
    fecha_completado: Optional[str] = None
    proyecto: Optional[str] = None
    anio: Optional[str] = None               # TEXT en la base, "2026"
    mes: Optional[str] = None                # tal cual la celda ("marzo"/"Marzo")
    # Atributos del catálogo que esta fila aporta si su capacitación hay que crearla.
    tipo: Optional[str] = None
    entidad_capacitadora: Optional[str] = None
    modalidad: Optional[str] = None
    duracion_horas: Optional[float] = None
    avisos: List[str] = []                   # p. ej. "sin fecha derivable", "ambiguo en el padrón"


class CapacitacionACrear(BaseModel):
    """Un título que no existe en el catálogo de la empresa y el confirmar va a crear."""
    nombre: str
    tipo: Optional[str] = None
    entidad_capacitadora: Optional[str] = None
    modalidad: Optional[str] = None
    duracion_horas: Optional[float] = None
    avisos: List[str] = []                   # p. ej. la duración conflictiva entre filas


class ParParecido(BaseModel):
    """Dos nombres crudos distintos que probablemente sean la misma persona. RRHH decide."""
    nombre_a: str
    nombre_b: str
    motivo: str


class ImportacionFormacionPreviewResponse(BaseModel):
    filas_validas: List[FilaFormacionPreview]
    errores: List[FilaFormacionError]
    capacitaciones_a_crear: List[CapacitacionACrear]
    sin_match: List[str]                     # nombres que van a entrar como nombre_libre
    pares_parecidos: List[ParParecido]
    hoja_leida: Optional[str] = None
    total_hojas: int = 1


class ImportacionFormacionConfirmarRequest(BaseModel):
    empresa_id: UUID
    filas: List[FilaFormacionPreview]


class ImportacionFormacionConfirmarResponse(BaseModel):
    importados: int
    errores: List[FilaFormacionError]
    capacitaciones_creadas: List[str]        # nombres creados en el catálogo por este lote
